contracts/Bridge/BaseXVSProxyOFT.sol

```diff
@@ -0,0 +1,356 @@
+     */
+    event SetMaxSingleTransactionLimit(uint16 chainId, uint256 oldMaxLimit, uint256 newMaxLimit);
+    /**
+     * @notice Emitted when the maximum daily limit of transactions from local chain is modified.
+     */
+    event SetMaxDailyLimit(uint16 chainId, uint256 oldMaxLimit, uint256 newMaxLimit);
+    /**
+     * @notice Emitted when the maximum limit for a single receive transaction from remote chain is modified.
+     */
+    event SetMaxSingleReceiveTransactionLimit(uint16 chainId, uint256 oldMaxLimit, uint256 newMaxLimit);
+    /**
+     * @notice Emitted when the maximum daily limit for receiving transactions from remote chain is modified.
+     */
+    event SetMaxDailyReceiveLimit(uint16 chainId, uint256 oldMaxLimit, uint256 newMaxLimit);
+    /**
+     * @notice Event emitted when oracle is modified.
+     */
+    event OracleChanged(address indexed oldOracle, address indexed newOracle);
+    /**
+     * @notice Event emitted when trusted remote sets to empty.
+     */
+    event TrustedRemoteRemoved(uint16 chainId);
+
+    /**
+     * @param tokenAddress_ Address of the inner token.
+     * @param sharedDecimals_ No of shared decimals.
+     * @param lzEndpoint_ Address of the layer zero endpoint contract.
+     * @param oracle_ Address of the price oracle.
+     * @custom:error ZeroAddressNotAllowed is thrown when token contract address is zero.
+     * @custom:error ZeroAddressNotAllowed is thrown when lzEndpoint contract address is zero.
+     * @custom:error ZeroAddressNotAllowed is thrown when oracle contract address is zero.
+     */
+    constructor(
+        address tokenAddress_,
+        uint8 sharedDecimals_,
+        address lzEndpoint_,
+        address oracle_
+    ) BaseOFTV2(sharedDecimals_, lzEndpoint_) {
+        ensureNonzeroAddress(tokenAddress_);
+        ensureNonzeroAddress(lzEndpoint_);
+
+        innerToken = IERC20(tokenAddress_);
+
+        (bool success, bytes memory data) = tokenAddress_.staticcall(abi.encodeWithSignature("decimals()"));
+        require(success, "ProxyOFT: failed to get token decimals");
+        uint8 decimals = abi.decode(data, (uint8));
+
+        require(sharedDecimals_ <= decimals, "ProxyOFT: sharedDecimals must be <= decimals");
+        ld2sdRate = 10 ** (decimals - sharedDecimals_);
+
+        ensureNonzeroAddress(oracle_);
+        oracle = ResilientOracleInterface(oracle_);
+    }
+
+    /**
+     * @notice Set the address of the ResilientOracle contract.
+     * @dev Reverts if the new address is zero.
+     * @param oracleAddress_ The new address of the ResilientOracle contract.
+     * @custom:access Only owner.
+     * @custom:event Emits OracleChanged with old and new oracle address.
+     */
+    function setOracle(address oracleAddress_) external onlyOwner {
+        ensureNonzeroAddress(oracleAddress_);
+        emit OracleChanged(address(oracle), oracleAddress_);
```

Comment:
nit: I instinctively think about emitting events last although technically there isn't a difference with how it is currently written